### PR TITLE
Visualizer for Pipelined CPU's

### PR DESCRIPTION
**Pull request by klanmiko**
This visualizer was written in scala-swing and the goal is to display the modules used in the pipelined CPU design and their relative location between the pipeline registers. The visualization should reflect the value of I/O on the modules and update every cycle.

Features:
 - Designed with an MVC architecture based on a simplified circuit module and description of the pipelined CPU.
 - Loads LowFIRRTL through some AST parsing
 - Automatically places instances of modules to their correct pipeline register by running a DFS on the node graph to find modules connected to each pipeline register
 - Uses Treadle's peek() call to load values at ports.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -92,7 +95,7 @@ lazy val root = (project in file("."))
     libraryDependencies += scalatest % Lab4,
     testOptions in TestAll := Seq(Tests.Filter(allFilter)),
     // CHANGE THE LINE BELOW FOR EACH LAB!!!! Use the matching filter
-    testOptions in Test := Seq(Tests.Filter(allFilter)),
+    testOptions in Test := Seq(Tests.Filter(lab3Filter)),
```
Comment (powerjg) [inline build.sbt]: This change should be reverted :)

**Review by powerjg — COMMENTED** [review]
*Very* cool! I'm super excited to use this next quarter!

You should make sure your name is attached to this with more than just the commit message. I would suggest adding an "about" dialog or putting "By Kaelan..." in one of the bottom corners.

**Review by powerjg — COMMENTED** [review]
Looks like a couple of files got added accidentally on this commit :).

**Comment by powerjg** [comment]
I just tried to run this and got the following error:

```
> singularity exec library://jlowepower/default/dinocpu sbt "runMain dinocpu.visualize multiply.riscv" 
java.lang.UnsatisfiedLinkError: /usr/local/openjdk-8/jre/lib/amd64/libawt_xawt.so: libXext.so.6: cannot open shared object file: No such file or directory
```

It looks like I might need to update the singularity container to have some other dependency installed. Do you know what the dependencies are for this?

**Comment by powerjg** [comment]
I figured it out. The following needs to be added to the `dinocpu.def`:

```
%post
    curl -L -o sbt-1.3.4.deb https://dl.bintray.com/sbt/debian/sbt-1.3.4.deb && \
    dpkg -i sbt-1.3.4.deb && \
    rm sbt-1.3.4.deb && \
    apt-get update -y && \
    apt-get install sbt libxext6 libfontconfig1 libxrender1 libxtst6 -y

```

**Comment by klanmiko** [comment]
> I figured it out. The following needs to be added to the `dinocpu.def`:

Ah. I never tested in the singularity container as I have sbt on my local development environment. I guess the old container didn't have awt and dependencies.

**Review by powerjg — APPROVED** [review]
LGTM! There's a lot that could be improved, but it's a really solid addition. Thanks!

**Pull request merged**